#AQI Prediction Model using Python

- PM2.5 PM10
- NO, NO2
- NH3-Ammonia
- CO
- So2
- O3
- Benzene, Toluene, Xylene

In [2]:
#Importing necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from warnings import filterwarnings
filterwarnings('ignore')

In [4]:
#there are two ways to see the data
#1.a "Series" is a one-dimensional labeled array, essentially representing a single column of data,
#while a "DataFrame" is a two-dimensional data structure with rows and columns
df = pd.read_csv('/content/air quality data.csv')


In [5]:
df.head() #top 5 rows

City        Date  PM2.5  PM10     NO    NO2    NOx  NH3     CO    SO2  \
0  Ahmedabad  2015-01-01    NaN   NaN   0.92  18.22  17.15  NaN   0.92  27.64   
1  Ahmedabad  2015-01-02    NaN   NaN   0.97  15.69  16.46  NaN   0.97  24.55   
2  Ahmedabad  2015-01-03    NaN   NaN  17.40  19.30  29.70  NaN  17.40  29.07   
3  Ahmedabad  2015-01-04    NaN   NaN   1.70  18.48  17.97  NaN   1.70  18.59   
4  Ahmedabad  2015-01-05    NaN   NaN  22.10  21.42  37.76  NaN  22.10  39.33   

       O3  Benzene  Toluene  Xylene  AQI AQI_Bucket  
0  133.36     0.00     0.02    0.00  NaN        NaN  
1   34.06     3.68     5.50    3.77  NaN        NaN  
2   30.70     6.80    16.40    2.25  NaN        NaN  
3   36.08     4.43    10.14    1.00  NaN        NaN  
4   39.31     7.01    18.89    2.78  NaN        NaN

In [6]:
df.shape
#shape- rows and columns

(29531, 16)

In [7]:
#information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29531 entries, 0 to 29530
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   City        29531 non-null  object 
 1   Date        29531 non-null  object 
 2   PM2.5       24933 non-null  float64
 3   PM10        18391 non-null  float64
 4   NO          25949 non-null  float64
 5   NO2         25946 non-null  float64
 6   NOx         25346 non-null  float64
 7   NH3         19203 non-null  float64
 8   CO          27472 non-null  float64
 9   SO2         25677 non-null  float64
 10  O3          25509 non-null  float64
 11  Benzene     23908 non-null  float64
 12  Toluene     21490 non-null  float64
 13  Xylene      11422 non-null  float64
 14  AQI         24850 non-null  float64
 15  AQI_Bucket  24850 non-null  object 
dtypes: float64(13), object(3)
memory usage: 3.6+ MB


In [8]:
#to check duplicate values
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
29526    False
29527    False
29528    False
29529    False
29530    False
Length: 29531, dtype: bool

In [9]:
df.duplicated().sum()

0

In [10]:
#to check missing values
df.isnull()

City   Date  PM2.5   PM10     NO    NO2    NOx    NH3     CO    SO2  \
0      False  False   True   True  False  False  False   True  False  False   
1      False  False   True   True  False  False  False   True  False  False   
2      False  False   True   True  False  False  False   True  False  False   
3      False  False   True   True  False  False  False   True  False  False   
4      False  False   True   True  False  False  False   True  False  False   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
29526  False  False  False  False  False  False  False  False  False  False   
29527  False  False  False  False  False  False  False  False  False  False   
29528  False  False  False  False  False  False  False  False  False  False   
29529  False  False  False  False  False  False  False  False  False  False   
29530  False  False  False  False  False  False  False  False  False  False   

          O3  Benzene  Toluene  Xylene    AQI  AQI_Bucket  
0      False    False    False   False   True        True  
1      False    False    False   False   True        True  
2      False    False    False   False   True        True  
3      False    False    False   False   True        True  
4      False    False    False   False   True        True  
...      ...      ...      ...     ...    ...         ...  
29526  False    False    False   False  False       False  
29527  False    False    False   False  False       False  
29528  False    False    False   False  False       False  
29529  False    False    False   False  False       False  
29530  False     True     True    True  False       False  

[29531 rows x 16 columns]

In [11]:
df.isnull().sum()

City              0
Date              0
PM2.5          4598
PM10          11140
NO             3582
NO2            3585
NOx            4185
NH3           10328
CO             2059
SO2            3854
O3             4022
Benzene        5623
Toluene        8041
Xylene        18109
AQI            4681
AQI_Bucket     4681
dtype: int64

In [12]:
#drop the rows where 'AQI' has missing values
df.dropna(subset=['AQI'], inplace=True)

In [13]:
df.isnull().sum().sort_values(ascending=False)

Xylene        15372
PM10           7086
NH3            6536
Toluene        5826
Benzene        3535
NOx            1857
O3              807
PM2.5           678
SO2             605
CO              445
NO2             391
NO              387
City              0
Date              0
AQI               0
AQI_Bucket        0
dtype: int64

In [14]:
df.shape

(24850, 16)

In [15]:
#summary of statistics in the dataset
df.describe()

PM2.5          PM10            NO           NO2           NOx  \
count  24172.000000  17764.000000  24463.000000  24459.000000  22993.000000   
mean      67.476613    118.454435     17.622421     28.978391     32.289012   
std       63.075398     89.487976     22.421138     24.627054     30.712855   
min        0.040000      0.030000      0.030000      0.010000      0.000000   
25%       29.000000     56.777500      5.660000     11.940000     13.110000   
50%       48.785000     96.180000      9.910000     22.100000     23.680000   
75%       80.925000    150.182500     20.030000     38.240000     40.170000   
max      914.940000    917.080000    390.680000    362.210000    378.240000   

                NH3            CO           SO2            O3       Benzene  \
count  18314.000000  24405.000000  24245.000000  24043.000000  21315.000000   
mean      23.848366      2.345267     14.362933     34.912885      3.458668   
std       25.875981      7.075208     17.428693     21.724525     16.036020   
min        0.010000      0.000000      0.010000      0.010000      0.000000   
25%        8.960000      0.590000      5.730000     19.250000      0.230000   
50%       16.310000      0.930000      9.220000     31.250000      1.290000   
75%       30.360000      1.480000     15.140000     46.080000      3.340000   
max      352.890000    175.810000    186.080000    257.730000    455.030000   

            Toluene       Xylene           AQI  
count  19024.000000  9478.000000  24850.000000  
mean       9.525714     3.588683    166.463581  
std       20.881085     6.754324    140.696585  
min        0.000000     0.000000     13.000000  
25%        1.027500     0.390000     81.000000  
50%        3.575000     1.420000    118.000000  
75%       10.180000     4.120000    208.000000  
max      454.850000   170.370000   2049.000000

In [16]:
 df.describe().T

count        mean         std    min      25%      50%       75%  \
PM2.5    24172.0   67.476613   63.075398   0.04  29.0000   48.785   80.9250   
PM10     17764.0  118.454435   89.487976   0.03  56.7775   96.180  150.1825   
NO       24463.0   17.622421   22.421138   0.03   5.6600    9.910   20.0300   
NO2      24459.0   28.978391   24.627054   0.01  11.9400   22.100   38.2400   
NOx      22993.0   32.289012   30.712855   0.00  13.1100   23.680   40.1700   
NH3      18314.0   23.848366   25.875981   0.01   8.9600   16.310   30.3600   
CO       24405.0    2.345267    7.075208   0.00   0.5900    0.930    1.4800   
SO2      24245.0   14.362933   17.428693   0.01   5.7300    9.220   15.1400   
O3       24043.0   34.912885   21.724525   0.01  19.2500   31.250   46.0800   
Benzene  21315.0    3.458668   16.036020   0.00   0.2300    1.290    3.3400   
Toluene  19024.0    9.525714   20.881085   0.00   1.0275    3.575   10.1800   
Xylene    9478.0    3.588683    6.754324   0.00   0.3900    1.420    4.1200   
AQI      24850.0  166.463581  140.696585  13.00  81.0000  118.000  208.0000   

             max  
PM2.5     914.94  
PM10      917.08  
NO        390.68  
NO2       362.21  
NOx       378.24  
NH3       352.89  
CO        175.81  
SO2       186.08  
O3        257.73  
Benzene   455.03  
Toluene   454.85  
Xylene    170.37  
AQI      2049.00

In [20]:
#percentage of the null values
null_values_percentage = (df.isnull().sum()/df.isnull().count()*100).sort_values(ascending=False)
null_values_percentage

Xylene        61.859155
PM10          28.515091
NH3           26.301811
Toluene       23.444668
Benzene       14.225352
NOx            7.472837
O3             3.247485
PM2.5          2.728370
SO2            2.434608
CO             1.790744
NO2            1.573441
NO             1.557344
City           0.000000
Date           0.000000
AQI            0.000000
AQI_Bucket     0.000000
dtype: float64

####Key Considerations:
- Xylene has the highest percentage of - missing values of 61.86%
- PM10n and NH3 28 - 26%